File: leetcode_cpp/linked_list_cycle_II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if (head == NULL || head->next == NULL) {
            return NULL;
        }
        ListNode *slow = head -> next;
        ListNode *fast = head -> next -> next;
        ListNode *third = head;
        while (fast != NULL) {
            if (slow == fast) {
                while (third != slow) {
                    third = third -> next;
                    slow = slow -> next;
                }
                return slow;
            }
            if (fast -> next == NULL) {
                return NULL;
            }
            slow = slow -> next;
            fast = fast -> next;
        }
        return NULL;
    }
};

File: leetcode_python/linked_list_cycle_II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a list node
    def detectCycle(self, head):
        if head is None or head.next is None:
            return None
        slow = head.next
        fast = head.next.next
        while fast:
            if slow == fast:
                third = head
                while third != slow:
                    third = third.next
                    slow = slow.next
                return slow
            if fast.next is None:
                return None
            slow = slow.next
            fast = fast.next.next
        return None
